lib/Base/NumpyReader.cpp

```diff
@@ -0,0 +1,324 @@
+      ok = true;
+    }
+  }
+  if (std::is_integral<To>::value && std::is_signed<To>::value) {
+    if ((std::is_integral<Ti>::value && std::is_signed<Ti>::value &&
+         sizeof(Ti) <= sizeof(To)) ||
+        (std::is_integral<Ti>::value && std::is_unsigned<Ti>::value &&
+         sizeof(Ti) < sizeof(To))) {
+      ok = true;
+    }
+  }
+  if (std::is_integral<To>::value && std::is_unsigned<To>::value) {
+    if (std::is_integral<Ti>::value && std::is_unsigned<Ti>::value &&
+        sizeof(Ti) <= sizeof(To)) {
+      ok = true;
+    }
+  }
+  CHECK(ok) << "NPY loader: Conversion not allowed, input type doesn't fit "
+               "into output type";
+
+  for (size_t i = 0; i < num_vals; i++) {
+    out.push_back((To) * (Ti *)&in[i * elem_size]);
+  }
+}
+
+/// Read npy file \p filename and store data tensor into vector \data. Tensor
+/// values are converted into \p type and tensor shape is provided in \p shape.
+template <typename T>
+static void numpyReader(const std::string &filename, std::vector<dim_t> &shape,
+                        std::vector<T> &data) {
+  // Map tensor type string to an enum.
+  enum NpyType {
+    Int8Ty,
+    UInt8Ty,
+    Int16Ty,
+    UInt16Ty,
+    Int32Ty,
+    UInt32Ty,
+    Int64Ty,
+    UInt64Ty,
+    Float16Ty,
+    Float32Ty,
+    Float64Ty
+  };
+
+  std::map<std::string, NpyType> npyType = {
+      {"i1", Int8Ty},    {"u1", UInt8Ty},   {"i2", Int16Ty},  {"u2", UInt16Ty},
+      {"i4", Int32Ty},   {"u4", UInt32Ty},  {"i8", Int64Ty},  {"u8", UInt64Ty},
+      {"f2", Float16Ty}, {"f4", Float32Ty}, {"f8", Float64Ty}};
+
+  // Wrapper for string find() method, that checks for return value.
+  auto find_str = [](std::string &str, const char *name,
+                     size_t *loc = nullptr) {
+    size_t newloc = loc ? str.find(name, *loc) : str.find(name);
+    if (newloc == std::string::npos) {
+      LOG(FATAL) << "NPY loader: Couldn't find string " << name;
+    }
+    return newloc;
+  };
+
+  // Wrapper for string find_first_of method, that checks for return value.
+  auto find_first_of = [](std::string &str, const char *find, size_t loc) {
+    loc = str.find_first_of(find, loc);
+    if (loc == std::string::npos) {
+      LOG(FATAL) << "NPY loader: Couldn't find string " << find;
+    }
+    return loc;
+  };
+
+  // temp vars.
+  size_t loc, locend;
+
+  // get file content
+  std::ifstream fs(filename.data(), std::ifstream::binary);
+  CHECK(fs) << "NPY loader: Couldn't open file: " << filename << "\n";
+  fs.seekg(0, fs.end);
+  int len = fs.tellg();
+  CHECK(len > 0) << "NPY loader: File too short: " << filename;
+  fs.seekg(0, fs.beg);
+  char *buffer = new char[len];
+  fs.read(buffer, len);
+  CHECK(fs) << "NPY loader: Reading file failed: " << filename;
+  fs.close();
+
+  // magic
+  uint8_t MAGIC[] = {0x93, 0x4E, 0x55, 0x4D, 0x50, 0x59};
+  if (memcmp((uint8_t *)buffer, MAGIC, 6)) {
+    LOG(FATAL) << "NPY loader: Magic number not found";
+  }
```

Comment:
This is done here and later on in this file -- perhaps make this a single static helper to use in both places, and add a bit more description for it?